src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -145,143 +144,143 @@ protected String name() {
-
-            if (currentRules[newVerb] == newAnchorId) {
-                return;
-            }
+            ViewGroup parent = (ViewGroup) found;
+            SparseArray<View> idToChild = new SparseArray<View>();
 
-            if (mOriginalValues.containsKey(found)) {
-                ; // Cache exactly one
-            } else {
-                LayoutRule originalValue = new LayoutRule(newVerb, currentRules[newVerb]);
-                mOriginalValues.put(found, originalValue);
+            int count = parent.getChildCount();
+            for (int i = 0; i < count; i++) {
+                View child = parent.getChildAt(i);
+                int childId = child.getId();
+                if (childId > 0) {
+                    idToChild.put(childId, child);
+                }
             }
 
-            try {
-                setLayout(found, newVerb, newAnchorId);
-            } catch (CantVisitException e) {
-                cleanup();
-                mOnEditErrorListener.onError(e);
-            }
-        }
+            int size = mArgs.size();
+            for (int i = 0; i < size; i++) {
+                LayoutRule layoutRule = mArgs.get(i);
+                final View currentChild = idToChild.get(layoutRule.viewId);
 
-        private void setLayout(View target, int verb, int anchorId) throws CantVisitException {
-            RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)target.getLayoutParams();
-            params.addRule(verb, anchorId);
+                RelativeLayout.LayoutParams currentParams = (RelativeLayout.LayoutParams) currentChild.getLayoutParams();
+                final int[] currentRules = currentParams.getRules().clone();
 
-            final Set<Integer> rules;
-            if (mHorizontalRules.contains(verb)) {
-                rules = mHorizontalRules;
-            } else if (mVerticalRules.contains(verb)) {
-                rules = mVerticalRules;
-            } else {
-                rules = null;
-            }
+                if (currentRules[layoutRule.verb] == layoutRule.anchor) {
+                    continue;
+                }
 
-            if (rules != null && !verifyLayout(target, rules)) {
-                throw new CantVisitException("Circular dependency detected!", "circular_dependency", mName);
-            }
+                if (mOriginalValues.containsKey(currentChild)) {
+                    ; // Cache exactly one set of rules per child view
+                } else {
+                    mOriginalValues.put(currentChild, currentRules);
+                }
 
-            target.setLayoutParams(params);
-        }
+                currentParams.addRule(layoutRule.verb, layoutRule.anchor);
 
-        private boolean verifyLayout(View target, Set<Integer> rules) {
-            ViewGroup parent = (ViewGroup) target.getParent();
-            SparseArray<View> idToChild = new SparseArray<View>();
+                final Set<Integer> rules;
```

Comment:
3d28ecb

It now validates the rules for every single change. This is to prevent weird bugs, like the user changes their implementation and the old layout changes don't work anymore.